lightning/src/routing/scoring.rs

```diff
@@ -1446,29 +1423,70 @@ impl<G: Deref<Target = NetworkGraph<L>>, L: Deref, T: Time> ScoreUpdate for Prob
 				let capacity_msat = channel.effective_capacity().as_msat();
 				self.channel_liquidities
 					.entry(hop.short_channel_id)
-					.or_insert_with(ChannelLiquidity::new)
-					.as_directed_mut(source, &target, capacity_msat, self.decay_params)
-					.successful(amount_msat, format_args!("SCID {}, towards {:?}", hop.short_channel_id, target), &self.logger);
+					.or_insert_with(|| ChannelLiquidity::new(duration_since_epoch))
+					.as_directed_mut(source, &target, capacity_msat)
+					.successful(amount_msat, duration_since_epoch,
+						format_args!("SCID {}, towards {:?}", hop.short_channel_id, target), &self.logger);
 			} else {
 				log_debug!(self.logger, "Not able to learn for channel with SCID {} as we do not have graph info for it (likely a route-hint last-hop).",
 					hop.short_channel_id);
 			}
 		}
 	}
 
-	fn probe_failed(&mut self, path: &Path, short_channel_id: u64) {
-		self.payment_path_failed(path, short_channel_id)
+	fn probe_failed(&mut self, path: &Path, short_channel_id: u64, duration_since_epoch: Duration) {
+		self.payment_path_failed(path, short_channel_id, duration_since_epoch)
 	}
 
-	fn probe_successful(&mut self, path: &Path) {
-		self.payment_path_failed(path, u64::max_value())
+	fn probe_successful(&mut self, path: &Path, duration_since_epoch: Duration) {
+		self.payment_path_failed(path, u64::max_value(), duration_since_epoch)
+	}
+
+	fn decay_liquidity_certainty(&mut self, duration_since_epoch: Duration) {
+		let decay_params = self.decay_params;
+		self.channel_liquidities.retain(|_scid, liquidity| {
+			liquidity.min_liquidity_offset_msat =
+				liquidity.decayed_offset(liquidity.min_liquidity_offset_msat, duration_since_epoch, decay_params);
+			liquidity.max_liquidity_offset_msat =
+				liquidity.decayed_offset(liquidity.max_liquidity_offset_msat, duration_since_epoch, decay_params);
+			liquidity.last_updated = duration_since_epoch;
+
+			let elapsed_time =
+				duration_since_epoch.saturating_sub(liquidity.offset_history_last_updated);
+			if elapsed_time > decay_params.historical_no_updates_half_life {
+				let half_life = decay_params.historical_no_updates_half_life.as_secs_f64();
+				if half_life != 0.0 {
+					let divisor = powf64(2048.0, elapsed_time.as_secs_f64() / half_life) as u64;
+					for bucket in liquidity.min_liquidity_offset_history.buckets.iter_mut() {
+						*bucket = ((*bucket as u64) * 1024 / divisor) as u16;
+					}
+					for bucket in liquidity.max_liquidity_offset_history.buckets.iter_mut() {
+						*bucket = ((*bucket as u64) * 1024 / divisor) as u16;
+					}
+					liquidity.offset_history_last_updated = duration_since_epoch;
+				}
```

Comment:
IIUC, this means we'll decay partial half-lives but only after decaying one full half-life. Why bother with using `2048.0` and `1024` here if this is happening in the background?

Reply:
But why not do partial decays when less than one half-life has passed?

Reply:
I see, SGTM.